### PR TITLE
Etcd Retry Logic

### PR DESCRIPTION
**Pull request by ntk148v**
Close: #73

### CODE REVIEW

File: pkg/autohealer/manager.go [hunk 1]
```diff
@@ -185,17 +187,36 @@ func (hm *Manager) save() {
...
+			if err := watchResp.Err(); err != nil {
+				level.Error(hm.logger).Log("msg", "Error watching etcd cloud provider keys", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctxc, cancelc = hm.etcdcli.WatchContext()
+					hm.watchc = hm.etcdcli.Watch(ctxc, model.DefaultCloudPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue WATCH
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: the loop is `WATCH` itself. Labeling is not needed here.

File: pkg/autohealer/manager.go [hunk 2]
```diff
@@ -231,8 +252,18 @@ func (hm *Manager) Run(ctx context.Context) {
...
+			if err := watchResp.Err(); err != nil {
+				level.Error(hm.logger).Log("msg", "Error watching etcd healer keys", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctxh, cancelh = hm.etcdcli.WatchContext()
+					hm.watchh = hm.etcdcli.Watch(ctxh, model.DefaultHealerPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue WATCH
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: here

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -188,14 +194,48 @@ func (c *Cluster) Run(ctx context.Context, rc chan struct{}) {
...
+			if err := watchResp.Err(); err != nil {
+				level.Error(c.logger).Log("msg", "Error watching cluster state", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctx, cancel = c.etcdcli.WatchContext()
+					watch = c.etcdcli.Watch(ctx, model.DefaultCloudPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue WATCH
```
Comment (vtdat) [inline pkg/cluster/cluster.go]: and here

**Comment by ntk148v** [comment]
> commented

Done,  please review and test it.

**Comment by vtdat** [comment]
> Overall, ok. However, it'd be nice if you add a line of log for every retries.

Sorry, I've just realize, it'd already had logs

**Comment by vtdat** [comment]
> > Overall, ok. However, it'd be nice if you add a line of log for every retries.
> 
> Sorry, I've just realize, it'd already had logs

```
level=debug ts=2020-01-07T07:59:06.394Z caller=etcd.go:257 component="etcd wrapper" msg="retry execute" action=keep-alive-once err="context deadline exceeded" key=3802568051608398600 count=0
level=debug ts=2020-01-07T07:59:13.395Z caller=etcd.go:257 component="etcd wrapper" msg="retry execute" action=keep-alive-once err="context deadline exceeded" key=3802568051608398600 count=1
level=debug ts=2020-01-07T07:59:20.396Z caller=etcd.go:257 component="etcd wrapper" msg="retry execute" action=keep-alive-once err="context deadline exceeded" key=3802568051608398600 count=2
```

Could you make the `count` starting from 1? It'd be more reasonable.

**Comment by vtdat** [comment]
Tested OK.

**Comment by ntk148v** [comment]
> > > Overall, ok. However, it'd be nice if you add a line of log for every retries.
> > 
> > 
> > Sorry, I've just realize, it'd already had logs
> 
> ```
> level=debug ts=2020-01-07T07:59:06.394Z caller=etcd.go:257 component="etcd wrapper" msg="retry execute" action=keep-alive-once err="context deadline exceeded" key=3802568051608398600 count=0
> level=debug ts=2020-01-07T07:59:13.395Z caller=etcd.go:257 component="etcd wrapper" msg="retry execute" action=keep-alive-once err="context deadline exceeded" key=3802568051608398600 count=1
> level=debug ts=2020-01-07T07:59:20.396Z caller=etcd.go:257 component="etcd wrapper" msg="retry execute" action=keep-alive-once err="context deadline exceeded" key=3802568051608398600 count=2
> ```
> 
> Could you make the `count` starting from 1? It'd be more reasonable.

OK, done in the newest commit.

File: pkg/autohealer/manager.go [hunk 4]
```diff
@@ -185,17 +187,35 @@ func (hm *Manager) save() {
 	}
 }
 
-// Run start healer mamaner instance
-func (hm *Manager) Run(ctx context.Context) {
-	hm.watchc = hm.etcdcli.Watch(ctx, model.DefaultCloudPrefix, etcdv3.WithPrefix())
-	hm.watchh = hm.etcdcli.Watch(ctx, model.DefaultHealerPrefix, etcdv3.WithPrefix())
+// Run start healer mananer instance
+func (hm *Manager) Run() {
+	retryCount := 0
```
[the remark this answers is not in the document]
Comment (vtdat) [reply pkg/autohealer/manager.go]: @ntk148v this one?

File: pkg/autohealer/manager.go [hunk 5]
```diff
@@ -185,17 +187,35 @@ func (hm *Manager) save() {
...
+				level.Error(hm.logger).Log("msg", "Error watching etcd cloud provider keys", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctxc, cancelc = hm.etcdcli.WatchContext()
+					hm.watchc = hm.etcdcli.Watch(ctxc, model.DefaultCloudPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue
+				}
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: This should have log also, please `count` from 1.

File: pkg/autohealer/manager.go [hunk 6]
```diff
@@ -231,8 +251,18 @@ func (hm *Manager) Run(ctx context.Context) {
...
+				level.Error(hm.logger).Log("msg", "Error watching etcd healer keys", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctxh, cancelh = hm.etcdcli.WatchContext()
+					hm.watchh = hm.etcdcli.Watch(ctxh, model.DefaultHealerPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue
+				}
```
Comment (vtdat) [inline pkg/autohealer/manager.go]: here

File: pkg/autoscaler/manager.go [hunk 7]
```diff
@@ -87,15 +89,29 @@ func (m *Manager) Stop() {
...
+				level.Error(m.logger).Log("msg", "Error watching etcd scaler keys", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctx, cancel = m.etcdcli.WatchContext()
+					watch = m.etcdcli.Watch(ctx, model.DefaultScalerPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue
+				}
```
Comment (vtdat) [inline pkg/autoscaler/manager.go]: here

File: pkg/cluster/cluster.go [hunk 8]
```diff
@@ -188,14 +193,48 @@ func (c *Cluster) Run(ctx context.Context, rc chan struct{}) {
...
+				level.Error(c.logger).Log("msg", "Error watching cluster state", "err", err)
+				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					// Re-init watch channel
+					ctx, cancel = c.etcdcli.WatchContext()
+					watch = c.etcdcli.Watch(ctx, model.DefaultCloudPrefix, etcdv3.WithPrefix())
+					// Increase retry count
+					retryCount += 1
+					time.Sleep(common.DefaultEtcdtIntervalBetweenRetries)
+					continue
+				}
```
Comment (vtdat) [inline pkg/cluster/cluster.go]: this block

**Comment by vtdat** [comment]
more logs pls.

**Comment by ntk148v** [comment]
> more logs pls.

doNe

File: pkg/cluster/cluster.go [hunk 9]
```diff
@@ -226,6 +226,8 @@ func (c *Cluster) Run(rc chan struct{}) {
 			if err := watchResp.Err(); err != nil {
 				level.Error(c.logger).Log("msg", "Error watching cluster state", "err", err)
 				if err == rpctypes.ErrNoLeader && retryCount <= common.DefaultEtcdRetryCount {
+					level.Debug(c.logger).Log("msg", "retry execute", "action", "watch",
+						"err", err, "key", model.DefaultCloudPrefix, "count", retryCount)
```
Comment (vtdat) [inline pkg/cluster/cluster.go]: `DefaultCloudPrefix` => `DefaultClusterPrefix` pls

**Pull request merged**